### PR TITLE
feat(training): archetype-aware AI judge scoring with behavior-specific rubrics

### PR DESCRIPTION
**Pull request by revlentless**
## Summary

- **Archetype-aware scoring**: AI Judge now evaluates agents based on how well they embody their archetype's personality, not just financial performance. A "Degen" who yolos and loses can score higher than a conservative agent that does nothing.
- **Python RLAIF integration**: Added `BehaviorMetrics` dataclass and 12 archetype-specific behavior bonus functions with custom reward weights (e.g., Degen: 60% behavior, 15% PnL vs Trader: 10% behavior, 55% PnL).
- **TypeScript RULER integration**: `RulerScoringService` now fetches archetype from DB and uses archetype-specific rubrics when comparing trajectories in the same scenario.

## Type

- [x] Feature
- [ ] Bug fix
- [ ] Refactor / cleanup (no behavior change)
- [ ] Performance
- [ ] Infra / ops
- [ ] Docs
- [ ] Contracts / on-chain
- [ ] Other: …

## Context / Links

- PR #7: The Advanced Judge (Archetype Rubrics)
- Branch: `feat/archetype-scoring-rubrics`
- Goal: Make `aiJudgeReward` in the database archetype-aware for RLAIF training

## Scope (keep it focused)

- [x] This PR is focused on a single change/theme (not a catch‑all)
- [x] Drive‑by refactors are excluded or split into a separate PR
- [x] Non‑goals / follow‑ups are listed below (with links)

**Areas touched**
- [ ] Web UI (`apps/web`)
- [ ] Web API routes / SSE / A2A (`apps/web`)
- [ ] CLI (`apps/cli`)
- [ ] Docs site (`apps/docs`) / vendor docs (`docs/vendors/*`)
- [ ] Domain / game engine (`packages/engine`, `packages/core/*`)
- [ ] Agents / runtime / A2A / MCP (`packages/agents`, `packages/a2a`, `packages/mcp`)
- [ ] API infra (`packages/api`)
- [ ] DB (`packages/db`)
- [ ] Contracts / on-chain (`packages/contracts`)
- [ ] Shared types/utils (`packages/shared`)
- [x] Tests (`packages/testing`)
- [x] Other: Training package (`packages/training`)

**Non-goals / follow-ups**
- LLM-as-judge temperature tuning for archetype scoring
- A/B testing archetype weights in production

## Changes

**Python (`packages/training/python/src/training/`)**
- `rubric_loader.py`: Added `normalize_archetype()`, versioning (`RUBRICS_VERSION`), and content hashing (`get_rubric_hash`, `get_all_rubrics_hash`)
- `rewards.py`: Added `ARCHETYPE_REWARD_WEIGHTS` (13 archetypes), `BehaviorMetrics` dataclass, `calculate_archetype_behavior_bonus()` with 12 specialized functions, and `archetype_composite_reward()` main scorer
- `babylon_env.py`: Updated `_load_trajectories()` to fetch archetype from DB, added `_extract_behavior_metrics()`, refactored `_score_with_judge()` to use archetype-aware scoring
- `__init__.py`: Exported new functions

**TypeScript (`packages/training/src/`)**
- `rubrics/index.ts`: Added `normalizeArchetype()` function, updated `getRubric()`, `getPriorityMetrics()`, `hasCustomRubric()` to use it
- `training/RulerScoringService.ts`: Fetch `archetype` in queries, pass through `groupByScenario()` and `scoreGroup()`, use archetype-specific rubrics in `buildJudgePrompt()` when all trajectories share same archetype

**Tests (`packages/training/python/tests/`)**
- `test_archetype_scoring.py`: 64 comprehensive tests covering all archetype scoring logic, boundary conditions, edge cases, and cross-archetype comparisons

**Misc**
- `.gitignore`: Added `.venv/` for Python virtual environments

## Review guide

**Start here**
- `packages/training/python/src/training/rewards.py` - Core archetype weights and behavior bonus logic
- `packages/training/src/training/RulerScoringService.ts` - TypeScript RULER integration

**Risk**
- [ ] Low
- [x] Medium
- [ ] High

**Notes for reviewers**
- The `ARCHETYPE_REWARD_WEIGHTS` dict defines how much each component (PnL, format, reasoning, behavior) matters per archetype. These weights are tunable.
- When RULER groups contain mixed archetypes, we fall back to the default rubric. Single-archetype groups use that archetype's specific rubric.
- Behavior bonus functions are clamped to [-0.3, 0.5] range to prevent dominating the score.

## Test plan

**Commands run**
- [x] `bun run check` (Biome format)
- [x] `bun run typecheck`
- [x] `bun run lint`
- [x] `bun run build`
- [ ] `bun run test` (unit + integration)
- [ ] `bun run test:e2e` (if critical flows changed)
- [ ] `bun run contracts:test` (if contracts changed)
- [ ] `bun run db:check` (if DB schema/queries changed)

**Python tests**
- [x] `pytest tests/test_archetype_scoring.py -v` (64 tests passing)

**Manual verification**
1. Verified a Degen with losses (-$50) scores higher (~0.45) than a Trader with same losses (~0.15) due to behavior weighting
2. Verified Social Butterfly with zero PnL but high social engagement scores well (~0.55)

## Ops / Migration / Deployment

- [x] No deploy impact
- [ ] Requires env var updates (listed below + `.env.example` updated)
- [ ] Requires DB migration (`bun run db:migrate`) / backfill / seed
- [ ] Changes cron schedule or endpoints (`vercel.json`, `CRON_SECRET`, etc.)
- [ ] Rollout behind a flag / gradual rollout

## Breaking changes

- [x] None
- [ ] Yes (describe + migration guide below)

## Security / privacy

- [x] No security impact
- [ ] Needs extra review (describe)

## Screenshots / recordings (UI)

- N/A (backend/training changes only)

## Checklist (author)
- [x] Self-review done (diff + critical paths)
- [x] Base branch is correct (`staging` by default)
- [x] Handlers remain thin and portable (validate → service → map errors)
- [x] Domain logic stays in packages (no Next/React/Elysia coupling in core)
- [x] `.env.example` updated (if env changed) and variables documented above
- [ ] Docs updated (and `bun run docs:generate` if needed)
- [x] Tests added/updated for behavior changes (or rationale provided)
- [ ] Dependency changes are intentional (`bun.lock` updated)
- [ ] Code owners requested (auto via CODEOWNERS or manual)

## Summary by CodeRabbit

* **New Features**
  * Archetype-aware scoring: weighted composite rewards, behavior-metric integration, archetype normalization/whitelisting, and exposed archetype utilities.
  * Scoring flow, prompts, and rubric selection now include archetype context; priority metrics shown first in evaluation prompts.
  * Rubric versioning and content hashing with canonical archetype list.

* **Bug Fixes**
  * Robust handling for malformed trajectory payloads with warnings and safe skips.

* **Tests**
  * Large, comprehensive test suite covering archetype weights, behavior bonuses, composite rewards, rubric/versioning, and edge cases.

* **Chores**
  * Ignore Python virtual environment directory (.venv) in .gitignore.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: packages/training/src/rubrics/__tests__/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  });
+
+  it('should handle mixed case with underscores', () => {
+    expect(normalizeArchetype('Social_Butterfly')).toBe('social-butterfly');
+    expect(normalizeArchetype('PERPS_TRADER')).toBe('perps-trader');
+  });
+
+  it('should handle empty string', () => {
+    expect(normalizeArchetype('')).toBe('');
+  });
```
Comment (slkzgm) [inline packages/training/src/rubrics/__tests__/index.test.ts:45]: `normalizeArchetype('')` currently returns `default` (see implementation in `packages/training/src/rubrics/index.ts`). This expectation looks inconsistent and will likely fail. Should the test expect `default`, or should `normalizeArchetype` preserve empty strings?

File: packages/training/python/src/training/babylon_env.py [hunk 2]
```diff
@@ -628,6 +668,152 @@ async def _score_with_judge(self, rollout_data: List[Dict]) -> Optional[ScoredDa
...
+                    unique_markets.add(str(market_id))
+
+                # Track position size
+                size = params.get("amount") or params.get("size") or params.get("quantity")
+                if size:
+                    metrics.avg_position_size += float(size)
+
+            # Prediction actions
+            elif action_type in ("predict", "bet", "forecast"):
+                metrics.predictions_made += 1
```
Comment (slkzgm) [inline packages/training/python/src/training/babylon_env.py:728]: As written, this `elif` won’t run for `predict`/`bet` because those action types are already handled in the trading branch above. If you want `predictions_made` / `prediction_accuracy` for `super-predictor` bonuses, consider counting prediction actions separately (even when they’re also trades), or remove `predict`/`bet` from the first branch.

File: packages/training/python/src/training/babylon_env.py [hunk 3]
```diff
@@ -628,6 +668,152 @@ async def _score_with_judge(self, rollout_data: List[Dict]) -> Optional[ScoredDa
...
+
+            elif action_type in ("share_info", "share"):
+                metrics.info_shared += 1
+
+            # Track reputation from environment state (support both camelCase and snake_case)
+            # Only convert to int if the value exists and is not None
+            env_state = step.get("environmentState", step.get("environment_state", {}))
+            if "reputationDelta" in env_state and env_state["reputationDelta"] is not None:
+                metrics.reputation_delta = int(env_state["reputationDelta"])
+            elif "reputation_delta" in env_state and env_state["reputation_delta"] is not None:
```
Comment (slkzgm) [inline packages/training/python/src/training/babylon_env.py:793]: Question: are `reputationDelta` / `followersGained` / etc per-step deltas or cumulative totals? Right now we overwrite these fields each step, which keeps only the last value. If they’re deltas, I’d expect `+=` accumulation instead.

File: packages/training/python/src/training/rubric_loader.py [hunk 4]
```diff
@@ -76,6 +82,26 @@ def has_custom_rubric(self, archetype: str) -> bool:
...
+        rubric = self.get_rubric(archetype)
+        return hashlib.sha256(rubric.encode()).hexdigest()[:16]
+    
+    def get_all_rubrics_hash(self) -> str:
+        """
+        Get combined hash of all rubrics.
+        Used for detecting any rubric changes.
+        """
+        all_rubrics = "::".join(sorted(self._rubrics.values())) + self._default_rubric
+        return hashlib.sha256(all_rubrics.encode()).hexdigest()[:16]
```
Comment (slkzgm) [inline packages/training/python/src/training/rubric_loader.py:99]: `get_all_rubrics_hash()` hashes `sorted(self._rubrics.values())`, which drops key association and likely won’t match TS ordering. If cross-language equality matters, could we hash a stable `[(key, value)]` ordering (e.g., sort by key) or hash the raw JSON file contents instead?

File: packages/training/config/rubrics.json [hunk 5]
```diff
@@ -134,4 +134,4 @@
     "infosec",
     "liar"
   ]
-}
+}
```
Comment (slkzgm) [inline packages/training/config/rubrics.json]: Minor: this generated JSON now has no trailing newline (`\ No newline at end of file`). If we want clean diffs, consider having the exporter append a newline when writing.

File: packages/training/scripts/export-rubrics.ts [hunk 6]
```diff
@@ -40,5 +57,8 @@ if (!fs.existsSync(configDir)) {
 
 fs.writeFileSync(outputPath, JSON.stringify(exportData, null, 2));
 console.log(`✓ Exported rubrics to ${outputPath}`);
```
Comment (slkzgm) [inline packages/training/scripts/export-rubrics.ts:59]: Related: consider appending a newline when writing the file (e.g., `fs.writeFileSync(outputPath, JSON.stringify(exportData, null, 2) + '\n')`) so POSIX tools/lints don’t complain and diffs stay stable.

File: packages/training/src/rubrics/index.ts [hunk 7]
```diff
@@ -111,27 +111,81 @@ export const PRIORITY_METRICS: Record<string, string[]> = {
   liar: LIAR_PRIORITY_METRICS,
 };
 
+/**
+ * Valid canonical archetype names for whitelist validation
+ */
+export const VALID_ARCHETYPES = new Set([
+  'trader',
```
Comment (slkzgm) [inline packages/training/src/rubrics/index.ts]: Potential drift: `VALID_ARCHETYPES` duplicates the canonical list from `getAvailableArchetypes()`. Any chance to derive the Set from the canonical array to keep a single source of truth?

File: packages/training/src/rubrics/index.ts [hunk 8]
```diff
@@ -111,27 +111,81 @@ export const PRIORITY_METRICS: Record<string, string[]> = {
...
+  'perps-trader',
+  'super-predictor',
+  'infosec',
+  'liar',
+]);
+
+/**
+ * Normalize archetype string to canonical format (lowercase, hyphens)
+ * Returns 'default' for empty/null/invalid archetypes
+ */
```
Comment (slkzgm) [inline packages/training/src/rubrics/index.ts:124]: Doc nit: `normalizeArchetype()` normalizes case/underscores, but doesn’t validate unknown values (e.g., `'foo'` → `'foo'`). The comment mentioning “invalid archetypes” might be misleading unless you intend to fold validation into normalization (vs keeping it in `sanitizeArchetype`).

File: packages/training/src/scoring/JudgePromptBuilder.ts [hunk 9]
```diff
@@ -350,6 +360,84 @@ The metrics provided are CONTEXT to inform your judgment. Use them to understand
...
+  /**
+   * Get a metric value from the metrics object using a dot-path
+   */
+  private getMetricValue(metrics: BehavioralMetrics, path: string): string {
+    const [category, key] = path.split('.');
+    if (!category || !key) return 'N/A';
+
+    // Access nested metric value based on category
+    let value: unknown;
+    switch (category) {
```
Comment (slkzgm) [inline packages/training/src/scoring/JudgePromptBuilder.ts:372]: Small TS nit: `unknown` here may violate our internal guideline. Could this be typed as a narrow union (`string | number | boolean | null | undefined`) or a shared `BehavioralMetricValue` type given the known metric shapes?

File: packages/training/python/tests/test_archetype_scoring.py [hunk 10]
```diff
@@ -0,0 +1,1025 @@
...
+            final_pnl=500.0, starting_balance=10000.0, end_balance=10500.0,
+            format_score=0.7, reasoning_score=0.6,
+        )
+        trading_metrics = BehaviorMetrics(
+            trades_executed=10, win_rate=0.7, markets_traded=4, total_pnl=500.0,
+        )
+
+        trader_score = archetype_composite_reward(trading_inputs, "trader", trading_metrics)
+        degen_score = archetype_composite_reward(trading_inputs, "degen", trading_metrics)
+        social_score = archetype_composite_reward(trading_inputs, "social-butterfly", trading_metrics)
```
Comment (slkzgm) [inline packages/training/python/tests/test_archetype_scoring.py:831]: `degen_score` is computed but not used in assertions here. Either assert something about it (e.g., compare to `trader_score`) or drop the variable to keep the test intent clear.

**Review by slkzgm — CHANGES_REQUESTED** [review]
Thanks for the detailed write-up and the strong direction here — archetype-aware scoring + rubric sync across TS/Python feels like the right abstraction for RLAIF.

I’m marking this as **REQUEST_CHANGES** for a couple of correctness/testability issues:

## Blocking
1) **TS test mismatch**: `normalizeArchetype('')` in the new Bun tests contradicts the current implementation (returns `default`), so I expect the test suite to fail as-is.
2) **Prediction metrics likely not counted**: in `_extract_behavior_metrics`, `predict`/`bet` are handled in the trading branch, making the subsequent “Prediction actions” branch unreachable for those action types. That likely prevents `predictions_made` / `prediction_accuracy` from being populated for `super-predictor` behavior bonuses.
3) **Rubrics hash semantics**: Python `get_all_rubrics_hash()` uses sorted values and drops keys; TS hashing is order-dependent. If hash equality is intended for cross-language cache invalidation/repro, we should align the strategy (or explicitly document that the hashes are language-local).

## Non-blocking
- Trailing newline in generated `packages/training/config/rubrics.json` (and updating `export-rubrics.ts` to write one) to avoid noisy diffs.
- `VALID_ARCHETYPES` duplicates the canonical list (potential drift); deriving it from `getAvailableArchetypes()` would keep one source of truth.
- Minor type cleanup in `JudgePromptBuilder` (`unknown` → a narrow union).
- Minor Python test cleanup (unused `degen_score`).

I also saw the bot reviews recommending running `bun run build` / TS tests; could you confirm those pass on this branch once the above is addressed?

Happy to re-review after the blockers are fixed.

**Comment by wtfsayo** [comment]
## CI Fix Status

**Fixed in this update:**
- Rebased on staging
- Commented out unused `experiencePlugin` import (was causing TS6133 error)

**Still failing:** Unit tests for `GameFeedbackSchema` - uses `example.com` URLs which are now blocked by schema validation.

**Blocked by:** PR #680 (test file fixes) needs to merge first.

**Action needed:** After #680 merges, rebase this PR to pick up the test fixes.

**Comment by revlentless** [comment]
@slkzgm @wtfsayo seeing your feedback now. Will go through each tonight. Thanks guys!

File: packages/training/src/rubrics/__tests__/index.test.ts (hunk 1 above)
Comment (revlentless) [reply packages/training/src/rubrics/__tests__/index.test.ts:45]: Fixed test expectation in `packages/training/src/rubrics/__tests__/index.test.ts` to expect 'default' instead of ''. 

Also added comprehensive tests for null and undefined cases. Additionally discovered and fixed the same issue in Python tests (`test_archetype_scoring.py`).

File: packages/training/python/src/training/babylon_env.py (hunk 2 above)
Comment (revlentless) [reply packages/training/python/src/training/babylon_env.py:728]: - Removed "predict" and "bet" from the trading actions conditional. 
- Changed prediction actions from elif to if so they can be counted independently and also track P&L. 
- Predictions now properly populate predictions_made, correct_predictions, and prediction_accuracy metrics while still counting as trading actions for P&L tracking.

File: packages/training/python/src/training/rubric_loader.py (hunk 4 above)
Comment (revlentless) [reply packages/training/python/src/training/rubric_loader.py:99]: - Fixed TypeScript implementation to sort values before hashing (matching Python implementation)
- Added comment noting the sorting is for cross-language consistency
- Both implementations now produce identical hashes.

File: packages/training/config/rubrics.json (hunk 5 above)
Comment (revlentless) [reply packages/training/config/rubrics.json]: Nice catch. Updated export script to append `\n` to JSON output

File: packages/training/src/rubrics/index.ts (hunk 7 above)
Comment (revlentless) [reply packages/training/src/rubrics/index.ts]: Changed from hardcoded array to derive directly from Object.keys(RUBRICS). Single source of truth maintained.

File: packages/training/src/rubrics/index.ts (hunk 8 above)
Comment (revlentless) [reply packages/training/src/rubrics/index.ts:124]: Clarified documentation to note that normalization doesn't validate against whitelist, and added reference to sanitizeArchetype() for validation use cases.

File: packages/training/src/scoring/JudgePromptBuilder.ts (hunk 9 above)
Comment (revlentless) [reply packages/training/src/scoring/JudgePromptBuilder.ts:372]: Changed unknown to proper union type `number | string | string[] | undefined` based on actual metric shapes (BehaviorMetrics includes `actionTypesUsed: string[]` and `dominantActionType: string`).

File: packages/training/python/tests/test_archetype_scoring.py (hunk 10 above)
Comment (revlentless) [reply packages/training/python/tests/test_archetype_scoring.py:831]: Removed unused variable on line 830 to keep test intent clear.

File: packages/training/python/src/training/babylon_env.py (hunk 3 above)
Comment (revlentless) [reply packages/training/python/src/training/babylon_env.py:793]: - Added clarifying comment explaining our assumption that these are cumulative totals (like agentBalance/agentPnL) and we intentionally take the last step's value.
- Included a note that if they turn out to be per-step deltas, the code should be changed from = to +=.

**Comment by SYMBaiEX** [comment]
I made a minor change after a quick review, was this PR ready for full review? I saw where you mentioned making adjustments tonight and didn't want to review it too early @revlentless

**Comment by revlentless** [comment]
> I made a minor change after a quick review, was this PR ready for full review? I saw where you mentioned making adjustments tonight and didn't want to review it too early @revlentless

i pushed up my changes, based on @slkzgm 's review, in this commit: 9f0b0da
Ready for review!

**Review by SYMBaiEX — APPROVED** [review]
## Ready for Merge ✅

### Action Needed
All feedback was addressed in commit 9f0b0da:
  - Fixed `normalizeArchetype('')` test expectation
  - Fixed prediction metrics counting (changed elif → if)  
  - Fixed hash ordering for cross-language consistency
  - Added trailing newline to JSON export
  - Made VALID_ARCHETYPES derive from RUBRICS keys
  - Fixed `unknown` type to `number | string | string[] | undefined`
  - Removed unused `degen_score` variable

The archetype-aware scoring implementation is solid with comprehensive test coverage. Ready to ship! 🚀
